File: frontend/src/types/workflow.ts
export type FieldType = 
  | 'str' 
  | 'int' 
  | 'bool' 
  | 'float' 
  | 'list[str]' 
  | 'list[int]' 
  | 'dict' 
  | 'list[dict[str, Any]]'
  | 'Any';

export interface SignatureField {
  name: string;
  type: FieldType;
  description?: string;
  required: boolean;
}

export interface NodePosition {
  x: number;
  y: number;
}

export type ModuleType = 
  | 'Predict' 
  | 'ChainOfThought' 
  | 'ReAct' 
  | 'BestOfN' 
  | 'Refine';

export type RetrieverType = 'UnstructuredRetrieve' | 'StructuredRetrieve';

export type LogicType = 'IfElse' | 'Merge' | 'FieldSelector';

export type NodeType = 'signature_field' | 'module' | 'logic' | 'retriever';

export interface BaseNodeData {
  label?: string;
}

export interface SignatureFieldNodeData extends BaseNodeData {
  fields: SignatureField[];
  isStart?: boolean;
  isEnd?: boolean;
  connectionMode?: 'whole' | 'field-level'; // Toggle between connection modes
}

export interface ModuleNodeData extends BaseNodeData {
  moduleType: ModuleType;
  model?: string;
  instruction?: string; // Task description for DSPy signature
  parameters: Record<string, any>;
}

export interface LogicNodeData extends BaseNodeData {
  logicType: LogicType;
  condition?: string;
  parameters: Record<string, any>;
  // FieldSelector specific data
  selectedFields?: string[];
  fieldMappings?: Record<string, string>; // old_name -> new_name
  availableFields?: SignatureField[]; // Fields available from input
}

export interface RetrieverNodeData extends BaseNodeData {
  retrieverType: RetrieverType;
  // UnstructuredRetrieve specific fields
  catalogName?: string; // Mandatory for UnstructuredRetrieve
  schemaName?: string; // Mandatory for UnstructuredRetrieve
  indexName?: string; // Mandatory for UnstructuredRetrieve
  contentColumn?: string; // Mandatory for UnstructuredRetrieve
  idColumn?: string; // Mandatory for UnstructuredRetrieve
  embeddingModel?: string; // Optional for UnstructuredRetrieve
  queryType?: 'HYBRID' | 'ANN'; // Default HYBRID for UnstructuredRetrieve
  numResults?: number; // Default 3 for UnstructuredRetrieve
  scoreThreshold?: number; // Default 0.0 for UnstructuredRetrieve
  // StructuredRetrieve specific fields
  genieSpaceId?: string; // Mandatory for StructuredRetrieve
  parameters: Record<string, any>;
}

export interface WorkflowNode {
  id: string;
  type: NodeType;
  position: NodePosition;
  data: SignatureFieldNodeData | ModuleNodeData | LogicNodeData | RetrieverNodeData;
}

export interface WorkflowEdge {
  id: string;
  source: string;
  target: string;
  sourceHandle?: string;
  targetHandle?: string;
  type?: string;
}

export interface Workflow {
  id: string;
  name: string;
  description?: string;
  nodes: WorkflowNode[];
  edges: WorkflowEdge[];
  createdAt: string;
  updatedAt: string;
}

export interface ExecutionResult {
  executionId: string;
  status: 'pending' | 'running' | 'completed' | 'failed';
  result?: Record<string, any>;
  error?: string;
  traceId?: string;
}

File: backend/app/models/workflow.py
from pydantic import BaseModel, Field
from typing import Dict, List, Any, Optional, Union, Literal
from enum import Enum
from datetime import datetime


class FieldType(str, Enum):
    STRING = "str"
    INTEGER = "int"
    BOOLEAN = "bool"
    FLOAT = "float"
    LIST_STRING = "list[str]"
    LIST_INT = "list[int]"
    DICT = "dict"
    ANY = "Any"


class SignatureField(BaseModel):
    name: str
    type: FieldType
    description: Optional[str] = None
    required: bool = True


class NodePosition(BaseModel):
    x: float
    y: float


class ModuleType(str, Enum):
    PREDICT = "Predict"
    CHAIN_OF_THOUGHT = "ChainOfThought"
    REACT = "ReAct"
    BEST_OF_N = "BestOfN"
    REFINE = "Refine"


class RetrieverType(str, Enum):
    UNSTRUCTURED_RETRIEVE = "UnstructuredRetrieve"
    STRUCTURED_RETRIEVE = "StructuredRetrieve"


class LogicType(str, Enum):
    IF_ELSE = "IfElse"
    MERGE = "Merge"
    FIELD_SELECTOR = "FieldSelector"


class NodeType(str, Enum):
    SIGNATURE_FIELD = "signature_field"
    MODULE = "module"
    LOGIC = "logic"
    RETRIEVER = "retriever"


class BaseNode(BaseModel):
    id: str
    type: NodeType
    position: NodePosition
    data: Dict[str, Any]


class SignatureFieldNode(BaseNode):
    type: Literal[NodeType.SIGNATURE_FIELD] = NodeType.SIGNATURE_FIELD
    data: Dict[str, Any] = Field(default_factory=lambda: {
        "fields": [],
        "is_start": False,
        "is_end": False,
        "connection_mode": "whole"  # "whole" or "field-level"
    })


class ModuleNode(BaseNode):
    type: Literal[NodeType.MODULE] = NodeType.MODULE
    data: Dict[str, Any] = Field(default_factory=lambda: {
        "module_type": None,
        "model": None,
        "instruction": None,
        "parameters": {}
    })


class LogicNode(BaseNode):
    type: Literal[NodeType.LOGIC] = NodeType.LOGIC
    data: Dict[str, Any] = Field(default_factory=lambda: {
        "logic_type": None,
        "condition": None,
        "parameters": {},
        "selected_fields": [],
        "field_mappings": {},
        "available_fields": []
    })


class RetrieverNode(BaseNode):
    type: Literal[NodeType.RETRIEVER] = NodeType.RETRIEVER
    data: Dict[str, Any] = Field(default_factory=lambda: {
        "retriever_type": None,
        # UnstructuredRetrieve fields
        "catalog_name": None,
        "schema_name": None, 
        "index_name": None,
        "embedding_model": None,
        "query_type": "HYBRID",
        "num_results": 3,
        "score_threshold": 0.0,
        # StructuredRetrieve fields
        "genie_space_id": None,
        "parameters": {}
    })


class Edge(BaseModel):
    id: str
    source: str
    target: str
    sourceHandle: Optional[str] = None
    targetHandle: Optional[str] = None
    type: Optional[str] = "default"


class Workflow(BaseModel):
    id: str
    name: str
    description: Optional[str] = None
    nodes: List[Union[SignatureFieldNode, ModuleNode, LogicNode, RetrieverNode]]
    edges: List[Edge]
    created_at: datetime = Field(default_factory=datetime.now)
    updated_at: datetime = Field(default_factory=datetime.now)


class WorkflowExecution(BaseModel):
    workflow_id: str
    input_data: Dict[str, Any]
    execution_id: str
    status: Literal["pending", "running", "completed", "failed"]
    result: Optional[Dict[str, Any]] = None
    error: Optional[str] = None
    trace_id: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.now)


class WorkflowCreateRequest(BaseModel):
    name: str
    description: str = ""
    nodes: List[Dict[str, Any]] = []
    edges: List[Dict[str, Any]] = []


class WorkflowUpdateRequest(BaseModel):
    name: str = None
    description: str = None
    nodes: List[Dict[str, Any]] = None
    edges: List[Dict[str, Any]] = None


class ExecutionRequest(BaseModel):
    input_data: Dict[str, Any]


class PlaygroundExecutionRequest(BaseModel):
    workflow_id: Optional[str] = None  # Optional workflow ID for tracking
    workflow_ir: Dict[str, Any]  # Workflow IR containing nodes and edges
    question: str
    conversation_history: list = []  # List of previous conversation exchanges
